Make opening a non-existent interface on Linux return 'no such interface'

afshin-paydar opened the pull request:
- https://github.com/the-tcpdump-group/libpcap/issues/1538
- https://github.com/the-tcpdump-group/tcpdump/issues/1334

Fix interface existence check on Linux before privileged operations

  When attempting to open a non-existent network interface on Linux,
  libpcap would return PCAP_ERROR_PERM_DENIED instead of the correct
  PCAP_ERROR_NO_SUCH_DEVICE. This occurred because the privileged
  socket(PF_PACKET, SOCK_RAW, 0) call in setup_socket() would fail
  with permission errors before interface validation could occur.

  This fix adds an interface existence check using an unprivileged
  AF_INET socket and SIOCGIFINDEX ioctl before attempting to create
  the privileged packet socket. Non-existent interfaces now correctly
  return PCAP_ERROR_NO_SUCH_DEVICE.

  This resolves the issue where applications like tcpdump would report
  "Permission denied" instead of "No such device exists" for non-existent
  interfaces, breaking their fallback logic from interface names to indices.

  Note: BSD/macOS platforms have the same underlying issue where BPF device
  access fails with permission errors before interface validation. This
  should be addressed in a separate commit.

  Fixes: libpcap issue #1538
  Related: tcpdump issue #1334

```
./tcpdump -i nonexistent
----------------------------------------
tcpdump: nonexistent: No such device exists
(No such device exists)

./tcpdump -i eth999
----------------------------------------
tcpdump: eth999: No such device exists
(No such device exists)

./tcpdump -i 000000000000008
----------------------------------------
tcpdump: Invalid adapter index 8: only 7 interfaces found

./tcpdump -i 00000000000000073
----------------------------------------
tcpdump: Invalid adapter index 73: only 7 interfaces found
```

infrastation commented on pcap-linux.c:
```diff
@@ -5164,13 +5164,13 @@ iface_get_ts_types(const char *device, pcap_t *handle, char *ebuf)
+			 * There's nothing more to say, so clear the
+			 * error message.
 			 */
-			handle->tstamp_type_list = NULL;
```
Please note that this explanation does not match the code: if the modified `iface_get_ts_types()` returns `PCAP_ERROR_NO_SUCH_DEVICE` (which is defined to -5), the test for -1 is false because -1 != -5 and there is no immediate return.

To add to that, modifying the immediate return condition in `pcapint_create_interface()` to include `PCAP_ERROR_NO_SUCH_DEVICE` would break things. The only way for `pcapint_create_interface()` to indicate a failure is to return `NULL`. To indicate a specific reason for a failure in a reliable way, it takes to get a `PCAP_ERROR_NO_SUCH_DEVICE` from `pcap_activate_linux()`, which requires that `pcapint_create_interface()` does not fail, which requires that `iface_get_ts_types()` does not fail, hence the current arrangement to defer the failure on `ENODEV`.

Therefore please leave `iface_get_ts_types()` intact.

infrastation reviewed (changes requested):
As noted.

afshin-paydar commented on pcap-linux.c:
```diff
@@ -4812,6 +4824,50 @@ pcap_setfilter_linux(pcap_t *handle, struct bpf_program *filter)
+	if (strlen(device) >= sizeof(ifr.ifr_name)) {
+		ebuf[0] = '\0';
+		return PCAP_ERROR_NO_SUCH_DEVICE;
+	}
```
Correct, removed redundant check

afshin-paydar committed:
Linux: check whether the device exists before privileged operations

In pcap_activate_linux(), check whether the device exists, using
an unprivileged ioctl socket from get_if_ioctl_socket(), before
attempting any operation that requires privileges, so that
attempting to open a non-existent device consistently reports
PCAP_ERROR_NO_SUCH_DEVICE rather than PCAP_ERROR_PERM_DENIED
from the attempt to create a PF_PACKET socket without
CAP_NET_RAW.

Fixes #1538.
Also addresses the-tcpdump-group/tcpdump#1334, where a
zero-padded numeric interface argument behaved differently
depending on whether its length reached IFNAMSIZ.

guyharris commented on pcap-linux.c:
```diff
@@ -2439,6 +2439,30 @@ setup_socket(pcap_t *handle, int is_any_device)
 	int			err = 0;
 	struct packet_mreq	mr;
 
+	/*
```
I'd put the routine that does this check into a separate routine, and perhaps call it earlier in `pcap_activate_linux()`.

afipay replied:
Thanks for review, moved it into a separate routine and call it early in `pcap_activate_linux()`

guyharris replied:
Looks good.

guyharris commented:
Should the check whether a device exists be done before trying to bind it, or after the attempt to bind the device fails with a permission error?

There's a time-of-check/time-of-use issue for both of those, but, if interfaces are appearing or disappearing, there's a risk of an error being reported for a condition that's not true by the time the user sees the error no matter *what* the code does, so I'm not bothered by that.

The one advantage of doing it afterwards is that fewer system calls are made in the case where everything succeeds, i.e. if the program *does* have permission to capture. I'm not sure that's a big enough difference to worry about.

infrastation commented:
Make it work, make it right, make it fast?
